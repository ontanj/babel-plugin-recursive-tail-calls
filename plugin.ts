import { type NodePath, types } from "@babel/core";
import type {
  CallExpression,
  Expression,
  Function,
  Identifier,
} from "@babel/types";

type t = typeof types;

interface State {
  // true if we should apply the transformation for this function
  recursion: boolean;
  // loop label
  labelIdentifier: Identifier;
  // identifier of this function
  functionIdentifier: Identifier;
  // identifier of loop condition
  conditionIdentifier: Identifier;
  // path of this function
  functionPath: NodePath<Function>;
  // name and default value of function argument
  arguments: { identifier: Identifier; defaultValue: Expression }[];
}

export default function ({ types: t }: { types: t }) {
  const callExpVisitor = {
    CallExpression(this: State, path: NodePath<CallExpression>) {
      const callsItself =
        t.isIdentifier(path.node.callee) &&
        path.scope.bindingIdentifierEquals(
          path.node.callee.name,
          this.functionIdentifier,
        );
      const isLast = t.isReturnStatement(path.parent);
      const shouldOptimize = callsItself && isLast;

      if (!shouldOptimize) return;

      this.recursion = true;

      const args = this.arguments.map(({ identifier, defaultValue }, index) => {
        return {
          identifier,
          value: path.node.arguments[index] ?? defaultValue,
        };
      });

      const updateExpression = t.expressionStatement(
        t.assignmentExpression(
          "=",
          t.arrayPattern(args.map(({ identifier }) => identifier)),
          t.arrayExpression(args.map(({ value }) => value as Expression)),
        ),
      );

      // the parent is ReturnStatement
      path.parentPath.insertBefore(updateExpression);
      path.parentPath.insertBefore(
        t.expressionStatement(
          t.assignmentExpression(
            "=",
            this.conditionIdentifier,
            t.booleanLiteral(true),
          ),
        ),
      );
      path.parentPath.insertBefore(t.continueStatement(this.labelIdentifier));
      path.parentPath.remove();
    },

    Function(path: NodePath<Function>) {
      // skip nested functions
      path.skip();
    },
  };

  return {
    visitor: {
      Function(path: NodePath<Function>) {
        const functionIdentifier = getFunctionIdentifier(path, t);
        if (!functionIdentifier) return;

        const functionBody = path.get("body");
        // until we support ternary, we can't have expression body
        if (!functionBody.isBlockStatement()) return;

        const labelIdentifier =
          path.scope.generateUidIdentifier("tail-call-loop");

        const conditionIdentifier =
          path.scope.generateUidIdentifier("continue-recursion");

        let args: State["arguments"];

        try {
          args = path.node.params.map(
            (param: (typeof path.node.params)[number]) => {
              if (t.isIdentifier(param)) {
                return {
                  identifier: param,
                  defaultValue: t.identifier("undefined"),
                };
              } else if (
                t.isAssignmentPattern(param) &&
                t.isIdentifier(param.left)
              ) {
                return { identifier: param.left, defaultValue: param.right };
              }
              throw new Error("Unsupported param expression");
            },
          );
        } catch (e: unknown) {
          return;
        }

        const state: State = {
          recursion: false,
          labelIdentifier,
          functionIdentifier,
          conditionIdentifier,
          functionPath: path,
          arguments: args,
        };

        path.traverse(callExpVisitor, state);

        // abort if there is no recursion
        if (!state.recursion) return;

        const conditionDeclaration = t.variableDeclaration("let", [
          t.variableDeclarator(conditionIdentifier, t.booleanLiteral(true)),
        ]);

        // wrap function body in while loop
        const whileStatement = t.whileStatement(
          conditionIdentifier,
          functionBody.node,
        );
        // insert `condition = false` first in loop
        functionBody.unshiftContainer(
          "body",
          t.expressionStatement(
            t.assignmentExpression(
              "=",
              conditionIdentifier,
              t.booleanLiteral(false),
            ),
          ),
        );
        const labeledStatement = t.labeledStatement(
          labelIdentifier,
          whileStatement,
        );
        const blockStatement = t.blockStatement([
          conditionDeclaration,
          labeledStatement,
        ]);

        functionBody.replaceWith(blockStatement);
      },
    },
  };
}

function getFunctionIdentifier(functionPath: NodePath<Function>, t: t) {
  if (t.isFunctionDeclaration(functionPath.node)) {
    return functionPath.node.id;
  } else if (t.isArrowFunctionExpression(functionPath.node)) {
    if (
      t.isVariableDeclarator(functionPath.parent) &&
      t.isIdentifier(functionPath.parent.id) &&
      functionPath.scope.getBinding(functionPath.parent.id.name)?.constant
    ) {
      return functionPath.parent.id;
    }
  }
}
