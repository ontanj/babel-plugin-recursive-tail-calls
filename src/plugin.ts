import { type NodePath, types } from "@babel/core";
import {
  isArrowFunctionExpression,
  isFunctionDeclaration,
  isIdentifier,
  isVariableDeclarator,
  type Function,
} from "@babel/types";
import {
  callExpressionRewriter,
  type State as CallExpressionState,
} from "./callExpressionRewriter.js";
import { findRecursion } from "./tailRecursionFinder.js";

type t = typeof types;

export default function ({ types: t }: { types: t }) {
  return {
    visitor: {
      Function(path: NodePath<Function>) {
        const functionIdentifier = getFunctionIdentifier(path);
        if (!functionIdentifier) return;

        const functionBody = path.get("body");

        const labelIdentifier =
          path.scope.generateUidIdentifier("tail-call-loop");

        const conditionIdentifier =
          path.scope.generateUidIdentifier("continue-recursion");

        let args: CallExpressionState["arguments"];

        try {
          args = path.node.params.map(
            (param: (typeof path.node.params)[number]) => {
              if (t.isIdentifier(param)) {
                return {
                  identifier: param,
                  defaultValue: t.identifier("undefined"),
                };
              } else if (
                t.isAssignmentPattern(param) &&
                t.isIdentifier(param.left)
              ) {
                return { identifier: param.left, defaultValue: param.right };
              }
              throw new Error("Unsupported param expression");
            },
          );
        } catch (e: unknown) {
          return;
        }

        if (functionBody.isExpression()) {
          if (!findRecursion(functionBody, functionIdentifier)) return;
          functionBody.replaceWith(
            t.blockStatement([t.returnStatement(functionBody.node)]),
          );
        }
        // this should never happen but is added for adequate type hints
        if (!functionBody.isBlockStatement())
          throw new Error("Body is not block statement");

        const state: CallExpressionState = {
          recursion: false,
          labelIdentifier,
          functionIdentifier,
          conditionIdentifier,
          functionPath: path,
          arguments: args,
        };

        path.traverse(callExpressionRewriter, state);

        // abort if there is no recursion
        if (!state.recursion) return;

        const conditionDeclaration = t.variableDeclaration("let", [
          t.variableDeclarator(conditionIdentifier, t.booleanLiteral(true)),
        ]);

        // wrap function body in while loop
        const whileStatement = t.whileStatement(
          conditionIdentifier,
          functionBody.node,
        );
        // insert `condition = false` first in loop
        functionBody.unshiftContainer(
          "body",
          t.expressionStatement(
            t.assignmentExpression(
              "=",
              conditionIdentifier,
              t.booleanLiteral(false),
            ),
          ),
        );
        const labeledStatement = t.labeledStatement(
          labelIdentifier,
          whileStatement,
        );
        const blockStatement = t.blockStatement([
          conditionDeclaration,
          labeledStatement,
        ]);

        functionBody.replaceWith(blockStatement);
      },
    },
  };
}

function getFunctionIdentifier(functionPath: NodePath<Function>) {
  if (functionPath.isFunctionDeclaration()) {
    return functionPath.node.id;
  } else if (functionPath.isArrowFunctionExpression()) {
    if (
      isVariableDeclarator(functionPath.parent) &&
      isIdentifier(functionPath.parent.id) &&
      functionPath.scope.getBinding(functionPath.parent.id.name)?.constant
    ) {
      return functionPath.parent.id;
    }
  }
}
