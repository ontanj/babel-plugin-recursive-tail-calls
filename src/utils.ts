import { NodePath } from "@babel/core";
import { CallExpression, isIdentifier } from "@babel/types";
import { Identifier } from "@babel/types";

export function isRecCall(
  path: NodePath<CallExpression>,
  identifier: Identifier,
) {
  return (
    isIdentifier(path.node.callee) &&
    path.scope.bindingIdentifierEquals(path.node.callee.name, identifier)
  );
}
