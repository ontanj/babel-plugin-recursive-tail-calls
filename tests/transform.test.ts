import babel from "@babel/core";
import plugin from "../src/plugin.js";
import {
  recursion,
  nestedStatements,
  multipleReturns,
  noReturn,
  defaultParameter,
  spread,
  exceedingArgs,
  logicalAnd,
  logicalOr,
  nullishCoalescing,
  ternary,
  combinedLogicalTernary,
  wrapAnonymous1,
  wrapAnonymous2,
  wrapAnonymous3,
} from "../examples/transform.before.js";
import { mapToTestCases } from "./utils.js";

const cases = mapToTestCases({
  recursion,
  nestedStatements,
  multipleReturns,
  noReturn,
  defaultParameter,
  spread,
  exceedingArgs,
  logicalAnd,
  logicalOr,
  nullishCoalescing,
  ternary,
  combinedLogicalTernary,
  wrapAnonymous1,
  wrapAnonymous2,
  wrapAnonymous3,
});

describe("transform", () => {
  it.each(cases)("%s", (_, testCase) => {
    const t = babel.transform(testCase, { plugins: [plugin] });
    expect(t?.code).toMatchSnapshot();
  });
});
