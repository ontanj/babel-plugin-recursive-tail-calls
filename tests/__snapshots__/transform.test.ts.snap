// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform base 1`] = `
"function base(a) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a <= 0) {
      return a;
    }
    [a] = [a - 1];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform combinedLogicalTernary 1`] = `
"function combinedLogicalTernary(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    const _left = a > 0;
    if (_left) {
      if (a > 10) {
        const _left2 = b;
        if (_left2) {
          return _left2;
        } else {
          [a, b] = [a + 1, (b + 1) % 5];
          _continueRecursion = true;
          continue _tailCallLoop;
        }
      } else {
        [a, b] = [a + 1, (b + 2) % 2];
        _continueRecursion = true;
        continue _tailCallLoop;
      }
    } else {
      return _left;
    }
  }
}"
`;

exports[`transform defaultArgActive 1`] = `
"function defaultArgActive(a, b = 0) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a < b) return a;
    [a, b] = [a - 1, 0];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform defaultArgInactive 1`] = `
"function defaultArgInactive(a, b = 0) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a < b) return a;
    [a, b] = [a - 1, b + 1];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform exceedingArgs 1`] = `
"function exceedingArgs(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a > b) return a - b;
    [a, b] = [a + b, b, console.log("CALL")];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform logicalAnd 1`] = `
"function logicalAnd(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a > 100) {
      return b;
    }
    const _left = a;
    if (_left) {
      [a, b] = [a + 1, b + 1];
      _continueRecursion = true;
      continue _tailCallLoop;
    } else {
      return _left;
    }
  }
}"
`;

exports[`transform logicalOr 1`] = `
"function logicalOr(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a > 100) {
      return b;
    }
    const _left = a;
    if (_left) {
      return _left;
    } else {
      [a, b] = [a + 1, b + 1];
      _continueRecursion = true;
      continue _tailCallLoop;
    }
  }
}"
`;

exports[`transform multipleArgs 1`] = `
"function multipleArgs(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a < b) return a;
    [a, b] = [a - 1, b + 1];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform multipleReturns 1`] = `
"function multipleReturns(a, b, c) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a < 0) return a;
    if (b < 0) {
      const n = b + c;
      [a, b, c] = [n, b, c + 1];
      _continueRecursion = true;
      continue _tailCallLoop;
    }
    [a, b, c] = [a + 1, b + 1, c];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform nestedStatements 1`] = `
"function nestedStatements(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a > b) {
      if (a > 0) {
        [a, b] = [a - 1, b];
        _continueRecursion = true;
        continue _tailCallLoop;
      }
      return a - b;
    }
    return a + b;
  }
}"
`;

exports[`transform noReturn 1`] = `
"function noReturn(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a < 10) {
      [a, b] = [a + b, b];
      _continueRecursion = true;
      continue _tailCallLoop;
    } else if (a % 2 === 0) {
      return a;
    }
  }
}"
`;

exports[`transform nullishCoalescing 1`] = `
"function nullishCoalescing(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a !== null && a !== void 0) {
      return a;
    } else {
      [a, b] = [b > 100 ? b : null, b + 1];
      _continueRecursion = true;
      continue _tailCallLoop;
    }
  }
}"
`;

exports[`transform ternary 1`] = `
"function ternary(a, b, c) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a > 10) {
      return c;
    } else {
      [a, b, c] = [a + b, b, c + a];
      _continueRecursion = true;
      continue _tailCallLoop;
    }
  }
}"
`;

exports[`transform undefinedArg 1`] = `
"function undefinedArg(a, b) {
  let _continueRecursion = true;
  _tailCallLoop: while (_continueRecursion) {
    _continueRecursion = false;
    if (a < b) return a;
    [a, b] = [a - 1, undefined];
    _continueRecursion = true;
    continue _tailCallLoop;
  }
}"
`;

exports[`transform wrapAnonymous1 1`] = `
"function wrapAnonymous1() {
  const arrowFunction = a => {
    let _continueRecursion2 = true;
    _tailCallLoop2: while (_continueRecursion2) {
      _continueRecursion2 = false;
      if (a <= 0) return a;
      [a] = [a - 1];
      _continueRecursion2 = true;
      continue _tailCallLoop2;
    }
  };
}"
`;

exports[`transform wrapAnonymous2 1`] = `
"function wrapAnonymous2() {
  const arrowFunctionNoBody = a => {
    let _continueRecursion2 = true;
    _tailCallLoop2: while (_continueRecursion2) {
      _continueRecursion2 = false;
      if (a <= 0) {
        return a;
      } else {
        [a] = [a - 1];
        _continueRecursion2 = true;
        continue _tailCallLoop2;
      }
    }
  };
}"
`;

exports[`transform wrapAnonymous3 1`] = `
"function wrapAnonymous3() {
  const functionExpression = function (a) {
    let _continueRecursion2 = true;
    _tailCallLoop2: while (_continueRecursion2) {
      _continueRecursion2 = false;
      if (a <= 0) return a;
      [a] = [a - 1];
      _continueRecursion2 = true;
      continue _tailCallLoop2;
    }
  };
}"
`;
